File: python/data_structure/my_queue.py
import unittest
from abc import ABC, abstractmethod


class InnerQueue(ABC):
    @abstractmethod
    def is_empty(self) -> bool: ...

    @abstractmethod
    def clear(self): ...

    @abstractmethod
    def enqueue(self, item): ...

    @abstractmethod
    def dequeue(self): ...

    @abstractmethod
    def peek(self): ...


class MyQueue(InnerQueue):
    def __init__(self):
        self._front = None
        self._rear = None
        self._size = 0

    class Node:
        def __init__(self, item):
            self.data = item
            self.next = None

    def is_empty(self) -> bool:
        return self._size == 0

    def clear(self):
        self._front = self._rear = None
        self._size = 0

    def enqueue(self, item):
        new_node = self.Node(item)
        if self.is_empty():
            self._front = new_node
        else:
            self._rear.next = new_node
        self._rear = new_node
        self._size += 1

    def dequeue(self):
        if self.is_empty():
            print('Queue is Empty.')
        else:
            tmp = self._front.data
            self._front = self._front.next
            self._size -= 1
            return tmp

    def poll(self):
        return self.dequeue()

    def peek(self):
        if self.is_empty():
            print('Queue is Empty.')
        else:
            return self._front.data

    def size(self) -> int:
        return self._size


class MyQueueTest(unittest.TestCase):
    def test_empty(self):
        my_queue = MyQueue()
        self.assertEqual(my_queue.size(), 0)
        self.assertTrue(my_queue.is_empty())
        self.assertEqual(my_queue.dequeue(), None)
        self.assertEqual(my_queue.poll(), None)
        self.assertEqual(my_queue.peek(), None)

    def test_all(self):
        my_queue = MyQueue()
        my_queue.enqueue(1)
        my_queue.enqueue(2)
        my_queue.enqueue(3)
        my_queue.enqueue(4)
        my_queue.enqueue(5)
        my_queue.enqueue(6)

        self.assertEqual(my_queue.size(), 6)
        self.assertFalse(my_queue.is_empty())
        self.assertEqual(my_queue.dequeue(), 1)
        self.assertEqual(my_queue.poll(), 2)
        self.assertEqual(my_queue.peek(), 3)
        self.assertEqual(my_queue.size(), 4)

        while my_queue.size() != 0:
            print('dequeue(): ' + str(my_queue.dequeue()))

        self.assertTrue(my_queue.size() == 0)

        my_queue.enqueue(3)
        my_queue.enqueue(4)
        my_queue.enqueue(5)
        self.assertEqual(my_queue.size(), 3)
        my_queue.clear()
        self.assertTrue(my_queue.is_empty())
        self.assertEqual(my_queue.size(), 0)

File: java/dataStructure/MyQueue.java
package dataStructure;

interface InnerMyQueue<T> {
    boolean isEmpty();
    void clear();
    void enqueue(T item);
    T dequeue();
    T peek();
}

public class MyQueue<T> implements InnerMyQueue<T> {
    private Node<T> front;
    private Node<T> rear;
    private int size = 0;

    private static class Node<T> {
        private T data;
        private Node<T> next;

        Node(T item) {
            this.data = item;
        }
    }

    @Override
    public boolean isEmpty() {
        return this.front == null;
    }

    @Override
    public void clear() {
        front = rear = null;
        size = 0;
    }

    @Override
    public void enqueue(T item) {
        Node<T> newNode = new Node<>(item);
        if (isEmpty())
            front = newNode;
        else
            rear.next = newNode;
        rear = newNode;
        size++;
    }

    @Override
    public T dequeue() {
        if (isEmpty()) {
            System.out.println("Queue is Empty");
            throw new java.util.NoSuchElementException();
        } else {
            T tmp = front.data;
            front = front.next;
            size--;
            return tmp;
        }
    }

    @Override
    public T peek() {
        if (isEmpty()) {
            System.out.println("Queue is Empty");
            throw new java.util.NoSuchElementException();
        } else {
            return front.data;
        }
    }

    public int getSize() {
        return this.size;
    }

    public void printAll() {
        if (front == null) {
            System.out.println("Queue is Empty");
        } else {
            Node<T> tmp = front;
            System.out.print("[");
            for (int i=0; i<size; i++) {
                System.out.print(tmp.data + ", ");
                tmp = tmp.next;
            }
            System.out.println("]");
        }
    }

    public static void main(String[] args) {
        MyQueue<Integer> queue = new MyQueue<>();
        System.out.println(queue.getSize());
        System.out.println(queue.isEmpty());
        queue.printAll();
        System.out.println("==============================");

        // System.out.println(queue.peek());
        // System.out.println(queue.dequeue());
        queue.enqueue(10);
        queue.enqueue(20);
        queue.enqueue(30);
        queue.enqueue(40);
        queue.enqueue(50);
        queue.enqueue(60);
        queue.printAll();
        System.out.println(queue.getSize());
        System.out.println("==============================");

        System.out.println(queue.peek());
        System.out.println(queue.getSize());
        System.out.println(queue.dequeue());
        System.out.println(queue.getSize());
        System.out.println("==============================");

        System.out.println(queue.front.data);
        System.out.println(queue.rear.data);
        queue.enqueue(90);
        queue.printAll();
        System.out.println(queue.isEmpty());
        System.out.println("==============================");

        queue.clear();
        queue.printAll();
        System.out.println(queue.getSize());
        System.out.println(queue.isEmpty());
    }
}
